[CodeComplete] Properly handle `if`/`switch` expressions

hamishknight opened the pull request:
Run PreCheckFunctionBodyRequest to ensure we insert an implicit return for an `if`/`switch` if needed, and ensure we don't try and type-check an element in a SingleValueStmtExpr separately, as it should be type-checked as a whole by the constraint system. This ensures we can propagate a contextual type from outside an `if`/`switch` expression for code completion.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

ahoppen commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2519,6 +2526,11 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
       if (isa<TapExpr>(E))
         return Action::SkipChildren(E);
 
+      // Don't walk into SingleValueStmtExprs, they should be type-checked as
+      // a whole.
+      if (isa<SingleValueStmtExpr>(E))
+        return Action::SkipChildren(E);
+
```
Just for understanding: This is the main change, right?

But if the `if` expression contains multiple statements, I think we do want to walk into it, right? E.g when you have one of the following.

```swift
enum Foo: Error {
  case bar
}

func test(_ x: Int) {}

let a: Foo = if true {
  test(#^COMPLETE^#)
  throw Foo.bar
} else {
  Foo.bar
}
```

Or even worse

```swift
enum Foo: Error {
  case bar
}

func test(_ x: Int) {}

let a: Foo = if true {
  func localFunc() {
    test(#^COMPLETE^#)
  }
  throw Foo.bar
} else {
  Foo.bar
}
```

In those cases all we want to type-check is `test(#^COMPLETE^#)` and not the entire `if` expression.

hamishknight replied:
Yeah this is the main change, we could definitely make it more efficient yeah, we really only need to type-check as a whole if the completion is in a single-expression branch

ahoppen replied:
> and even then we could still potentially skip type-checking non-expression branches

I think we should be skipping unrelated conjunctions when type checking for completion. I worked on this for result builders and it could also apply to if expressions, I just never checked. 

---

Could you add tests for the cases above. I agree that it’s a very uncommon thing to define a function inside an if expression, so if we aren’t taking the most optimal path, that’s probably OK. but we should definitely make sure that we do produce results.

hamishknight replied:
> I think we should be skipping unrelated conjunctions when type checking for completion

Yeah, I saw that, though I think we can be even more aggressive in this case and prune entire branches. Updated to include that, and added the extra test cases.

ahoppen commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2599,11 +2611,6 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
         // thus the transform couldn't be applied. Perform code completion
         // pretending there was no result builder to recover.
       }
-    } else if (func->hasSingleExpressionBody() &&
-                func->getResultInterfaceType()->isVoid()) {
-       // The function returns void.  We don't need an explicit return, no matter
-       // what the type of the expression is.  Take the inserted return back out.
-      func->getBody()->setLastElement(func->getSingleExpressionBody());
```
Was this simply no longer needed or was it rendered unnecessary by one of your changes?

hamishknight replied:
This is now done by PreCheckFunctionBodyRequest

ahoppen replied:
Oh, I missed that. Sorry

ahoppen commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2641,7 +2648,61 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
+                                          /*implicit*/ true));
+      body = BraceStmt::create(ctx, body->getLBraceLoc(), Elts,
+                               body->getRBraceLoc(), body->isImplicit());
+      AFD->setBody(body, AbstractFunctionDecl::BodyKind::Parsed);
```
Why do we need to do `AFT->setBody` here? AFAICT we weren’t doing that in the old implementation of `TypeCheckFunctionBodyRequest::evaluate`.

hamishknight replied:
I was trying to make the request be consistent in how it modifies the body (i.e always modifying it in-place), but yeah I'm not sure it's really needed.

ahoppen replied:
Ah, I see. Makes sense 👍🏽

ahoppen commented on test/IDE/complete_if_switch_expr.swift:
```diff
@@ -0,0 +1,118 @@
+}
+
+// DOT:      Begin completions, 2 items
+// DOT-NEXT: Decl[EnumElement]/CurrNominal/Flair[ExprSpecific]/TypeRelation[Convertible]: e[#E#]; name=e
+// DOT-NEXT: Decl[EnumElement]/CurrNominal/Flair[ExprSpecific]/TypeRelation[Convertible]: f({#Int#})[#E#]; name=f()
```
Nit: I’d do `DOT-DAG` so this doesn’t break in case we flip the order of these two for some reason.

hamishknight replied:
Good catch!

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented on lib/Sema/CSStep.cpp:
```diff
@@ -883,11 +883,15 @@ bool ConjunctionStep::attempt(const ConjunctionElement &element) {
+    // Avoid doing this for SingleValueStmtExprs, because we can more eagerly
+    // prune branches in that case, which requires us to detect the code
+    // completion option while solving the conjunction.
     CS.Options -= ConstraintSystemFlags::ForCodeCompletion;
```
@ahoppen Out of curiosity, is this even still needed now that we have 6cec68e302477f30264387a1b0533214aa5d6f8e? Seems like removing this could allow us to skip conjunction elements in more cases.

ahoppen replied:
Oh, I think we might be able to remove it. Would you mind opening a PR removing it and running the SourceKit stress tester? I’ll check the stress tester performance with the change.

hamishknight replied:
Sure thing: #67626

hamishknight commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2519,6 +2526,20 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
+      // SingleValueStmtExpr, don't walk into it. We'll want to type-check it
+      // a part of the whole expression.
+      if (auto *SVE = dyn_cast<SingleValueStmtExpr>(E)) {
+        SmallVector<Expr *> scratch;
+        for (auto *branch : SVE->getSingleExprBranches(scratch)) {
+          auto branchCharRange = Lexer::getCharSourceRangeFromSourceRange(
+            SM, branch->getSourceRange());
+          if (branchCharRange.contains(Loc))
+            return Action::SkipChildren(E);
+        }
```
One case we could still improve here is if the there's an inner closure/SVE within the single-expression branch, as we could still potentially walk in and find something to type-check independently. I'm not sure such cases come up all that often tho.

hamishknight replied:
Actually I think it should be simple enough to handle, let me update.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -1063,6 +1063,15 @@ class SyntacticElementConstraintGenerator
+    if (cs.isForCodeCompletion() && !braceStmt->getSingleActiveExpression() &&
+        locator->isForSingleValueStmtConjunctionOrBrace() &&
+        !cs.containsIDEInspectionTarget(braceStmt)) {
+      return;
```
Could this filtering be done earlier while generate constraints for if/switch expression?

hamishknight replied:
How does this look?

hamishknight replied:
Yeah that's what I was thinking

hamishknight replied:
Okay final version 🤞

xedin replied:
Looks great, thank you!

xedin commented on lib/Sema/ConstraintLocator.cpp:
```diff
@@ -634,22 +634,55 @@ bool ConstraintLocator::isForMacroExpansion() const {
+    return false;
+
+  auto path = getPath();
+  while (true) {
```
Extra-Nit: `while (!path.empty())` might be simpler?

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
Hmm GitHub seems bugged, current HEAD should be d521a7c395c02294739a830874128e521274e6ef

hamishknight commented:
@swift-ci please test

ahoppen commented:
@swift-ci Please SourceKit stress test

hamishknight committed:
[CodeComplete] More efficient skipping for completions in if/switch exprs

Skip type-checking multi-statement branches if the
completion is in a single-expression branch, and
skip type-checking the expression as a whole if
the completion is in a multi-statement branch.

hamishknight committed:
[CodeComplete] Query the Solution for contextual type

Instead of querying the ConstraintSystem, query the
Solution instead, as the ConstraintSystem is not
necessarily in the same state as when the solution
was produced.

hamishknight committed:
[Sema] Introduce PreCheckFunctionBodyRequest

Split out the part of TypeCheckFunctionBodyRequest
that inserts and removes implicit returns.

hamishknight committed:
[CodeComplete] Properly handle if/switch exprs

Run PreCheckFunctionBodyRequest to ensure we insert
an implicit return for an if/switch if needed, and
ensure we don't try and type-check an element in a
SingleValueStmtExpr separately, as it should be
type-checked as a whole by the constraint system.
This ensures we can propagate a contextual type from
outside an if/switch expression for code completion.

hamishknight committed:
[CodeComplete] Don't skip single-expression AssignExpr bodies of an if/switch expr

Fix an issue uncovered by the stress tester where
the brace element skipping logic could still attempt
to skip a single-expression body of an if/switch
expr.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
Stress tester uncovered an issue, fixed in latest commit
